### ISSUE TITLE
New cop: enforce positive ifs

### ISSUE THREAD

**Issue by Ana06** [rubocop-hq/rubocop#6129]
I would like to introduce a new IfPositive cop, which enforces to have positive conditions, `present?` instead or `empty?` for example, as they are equivalent just exchanging the if-else branches code and it is more readable as it is easier to think in positive. For example:

```ruby
# bad
if name.empty?
  @show_table = false
  @name = "undefined"
else
  @name = name
end

# good 
if name.present?
  @name = name
else
  @show_table = false
  @name = "undefined"
end
```

```ruby
#bad
realname.empty? ? user.login : "#{realname} (#{user.login})"

#good
realname.present? ? "#{realname} (#{user.login})" : user.login 
```

**Comment by bbatsov** [comment]
It's a good idea, but we either need to have some mapping of positive to negative mappings in the config of the cop or move this to `rubocop-rails`, if we're going to hardcode some Rails specific mappings there.

**Comment by rrosenblum** [comment]
There are some similar cops that already exist to try to enforce positive forms [Style/NegatedIf](https://github.com/rubocop-hq/rubocop/blob/master/manual/cops_style.md#stylenegatedif), [Style/InverseMethods](https://github.com/rubocop-hq/rubocop/blob/master/manual/cops_style.md#styleinversemethods), [Rails/Blank](https://github.com/rubocop-hq/rubocop/blob/master/manual/cops_rails.md#railsblank), and [Rails/Present](https://github.com/rubocop-hq/rubocop/blob/master/manual/cops_rails.md#railspresent).

None of these cops cover the case of converting `something.empty?` to `something.blank?` or `something.present?` because `empty?` and `blank?` are not direct synonyms for each other. Although, I guess if the code is checking `empty?` it implies that the object won't be `nil` so `blank?` or `present?` would be safe to use (unlike checking `nil?`). 

A few people commented on the work we did with safe navigation that they don't like safe navigation chaining because it can change implied behavior of code. If you know something can/will never be `nil`, don't use safe navigation because it implies the code can return `nil`. Because of this, I wouldn't want to convert `empty?` to `blank?` in all cases. Maybe there could be a configuration on `Rails/Blank` and `Rails/Present` to convert calls to `empty?`.

**Comment by tvw** [comment]
This is a very poor idea.

```ruby
"Some string".present?
```

gives me a `NoMethodError (undefined method `present?' for "":String)` in Ruby 2.5.1

while

```ruby
"Some string".empty?
```

does not.

Such a cop produces therefore tons of messages, which the developer can only stop by disabling the cop. The more such cops exist, the more cops must be disabled.

I think it is not a good idea to pollute rubocop with bad cops, because it encourages to disable cops as soon as possible.

And when I look at your examples: the ifs are "positive" as long as you do not use "not / !". If you have to cover both cases anywas your cop only makes a difference regarding the order of the code.

And last but not least: what is a better choice depends on the context. We already have a cop, which enforces the use of `empty?`. `empty?` is an alternative to `length == 0`. But it is not always a better alternative to `length == 0`. `empty?` refers to the state of a container, while `length` does this too, but the comparison, which is needed can also be seen, that it refers more to the elements of the container than the container itself.

Example:

I have in my project a container "@seen = []". As the name of the variable indicates, it is less important, that the variable is a container compared to the elements, it can contain.

So 

```ruby
if @seen.length == 0
```

is better than

```ruby
if @seen.empty?
```

because I am not interested in the state of the container, because length serves my as an indicator of how many items have been seen. 0 therefore is a better answer to the question "How many?" than "The container is empty". Because from the latter one you have to make an extra step: "When the container is empty it means, that no items have been seen." This is complicated.

The `empty?` method exists for contexts, where `empty?` is better than `length==0`, which means, where the state of the container is in the focus of the code rather than its elements. But the Empty-Cop cannot understand the context, nor does it care, if a `empty?`-method exists. It just jumps onto the usage of length and does not care, what length means. If you have a class "Line" where the line can have a length, when you test against zero, the cops encourages you to test against `empty?`. But the method does not exist for good reasons: a line cannot be empty or full.

Therefore the empty-cop is a bad cop, because it pollutes rubocops output with error messages and it encourages to write less readable and less understandable code.

**Comment by tvw** [comment]
> I interpret the rest of your comment as an argument that this cop should not be enabled by default.

I think, even in a rails app, the cop generates bogus messages. I understand the intent regarding ordinary rails models. But you can have much more than just rails models in your rails app. `present?` makes code more readable, where `present?` makes the code more understandable from a linguistic point of view. But even an empty string is a string and it is "present" and therefore confusing in some contexts, where length==0 is still better, eg. when counting things.

The cop cannot understand, what you are abstracting and which type of verbosity you need. You can see this also in her example: empty? is not the linguistic opposite of present?, but eg. absent? would be. The opposite of empty? is full? while there also is some state, which is neither empty? nor full?.

If you use "linguistic" method names, they should be used where it makes sense and not all over. But Rubocop encourages to use them all over, because it just spits out an error message which by the user can be seen as an recommendation for writing better code. In many cases the result will be the opposite: less readable code.

The reason for this is simple: for such a cop you need a decent amount of intelligence, because it must also understand the idea and the intent of your code and what you are abstracting about.

Then there is another issue:

```ruby
"".present? # => false
nil.present? # => false
"".empty? # => true
nil.empty? # => NoMethodError (undefined method `empty?' for nil:NilClass)
```

And in order to implement a "IfPositive cop" - as suggested, which enforces to have positive conditions, it should test against the usage of a "not (!)" over the whole condition. But even this could be complicated since it should only fire, if the code without "not" is easier to understand.

**Comment by mikegee** [comment]
> You can see this also in her example: empty? is not the linguistic opposite of present?

Indeed, a poor choice was made in the original examples here. `blank?` is the opposite of `present?` in ActiveSupport.
